envoy/stats/stats_macros.h

```diff
@@ -152,16 +152,41 @@ static inline std::string statPrefixJoin(absl::string_view prefix, absl::string_
  */
 #define MAKE_STATS_STRUCT(StatsStruct, StatNamesStruct, ALL_STATS)                                 \
   struct StatsStruct {                                                                             \
+    using StatNameType = StatNamesStruct;                                                          \
     StatsStruct(const StatNamesStruct& stat_names, Envoy::Stats::Scope& scope,                     \
                 Envoy::Stats::StatName prefix = Envoy::Stats::StatName())                          \
         : stat_names_(stat_names)                                                                  \
               ALL_STATS(MAKE_STATS_STRUCT_COUNTER_HELPER_, MAKE_STATS_STRUCT_GAUGE_HELPER_,        \
                         MAKE_STATS_STRUCT_HISTOGRAM_HELPER_,                                       \
                         MAKE_STATS_STRUCT_TEXT_READOUT_HELPER_,                                    \
                         MAKE_STATS_STRUCT_STATNAME_HELPER_) {}                                     \
-    const StatNamesStruct& stat_names_;                                                            \
+    const StatNameType& stat_names_;                                                               \
     ALL_STATS(GENERATE_COUNTER_STRUCT, GENERATE_GAUGE_STRUCT, GENERATE_HISTOGRAM_STRUCT,           \
               GENERATE_TEXT_READOUT_STRUCT, GENERATE_STATNAME_STRUCT)                              \
   }
 
+/**
+ * LazyInit$StatsStruct is a wrapper with creator of the actual "$StatsStruct"
+ * structure.
+ * It instantiate a $StatsStruct struct when any data member is referenced.
+ * See https://github.com/envoyproxy/envoy/issues/23575 for more details.
+ */
+template <typename StatsStructType> struct LazyInitStats {
+  LazyInitStats(Stats::Scope& scope, const typename StatsStructType::StatNameType& stat_names)
+      : scope_(scope), ctor_([&scope, &stat_names]() -> StatsStructType* {
```

Comment:
don't capture the ctor -- just capture the stat_names, as in https://github.com/envoyproxy/envoy/blob/c101df5598bacf02e3563938470e16c01bef9415/source/common/upstream/lazy_cluster_stats.h and pass the ctor to get (via shared inline helper)

The ctor you are creating here will be larger than the stat-names ref so you might as well be lazy in creating it.

Reply:
talked offline, I think we have reservation on performance implication of repeatedly creations of the ctor-lambda.

Reply:
Agreed. But then you can drop the scope member variable then I think. You are capturing it in the lambda anyway, and I think there are no other uses.

Reply:
this is done.